twilight-gateway/src/shard.rs

```diff
@@ -714,18 +764,55 @@ impl Shard {
         self.user_channel.sender()
     }
 
-    /// Close the shard's connection, providing a close frame indicating whether
-    /// a resume is intended.
+    /// Close the gateway connection with a close frame indicating whether to
+    /// also invalidate the shard's session.
+    ///
+    /// Returns the shard's session if the close frame code is not `1000` or
+    /// `1001`, which invalidates the session and shows the application's bot as
+    /// offline. Otherwise Discord will not invalidate the shard's session and
+    /// will continue to show the application's bot as online until its presence
+    /// times out.
+    ///
+    /// Sets status to [`ConnectionStatus::Disconnected`] with the `close_code`
+    /// from the `close_frame`.
+    ///
+    /// To read all remaining events, continue calling [`Shard::next_message`]
+    /// until it returns the response close message or a
+    /// [`ReceiveMessageErrorType::Io`] error type.
+    ///
+    /// You do not need to call this method upon receiving a close message to
+    /// respond do it, Twilight handles this for you.
     ///
-    /// Returns the gateway session of the shard, which may be provided via
-    /// [`ConfigBuilder::session`] to create a new shard that will resume the
-    /// gateway session.
+    /// # Example
     ///
-    /// If sending a close frame such as [`CloseFrame::NORMAL`] then Discord
-    /// will invalidate the shard's session, showing the application's bot as
-    /// offline. If sending a close frame such as [`CloseFrame::RESUME`] then
-    /// Discord will not invalidate the shard's session and will continue to
-    /// show the application's bot as online until its presence times out.
+    /// Close the gateway connection but process already received messages:
+    ///
+    /// ```no_run
+    /// # use twilight_gateway::{Intents, Shard, ShardId};
+    /// # #[tokio::main] async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let mut shard = Shard::new(ShardId::ONE, String::new(), Intents::empty());
+    /// use twilight_gateway::{
+    ///     error::ReceiveMessageErrorType,
+    ///     message::{CloseFrame, Message},
+    /// };
+    ///
+    /// shard.close(CloseFrame::NORMAL).await?;
+    ///
+    /// loop {
+    ///     match shard.next_message().await {
+    ///         Ok(Message::Close(_)) => {
+    ///             // We've now received a close message response from the
+    ///             // Gateway.
+    ///             // Further calls to `next_message` would cause a reconnect.
+    ///             break
+    ///         }
+    ///         Ok(_) => unimplemented!("handle message"),
```

Comment:
Would've liked to use `Ok(Message::Text(_))` but `Message` is marked as `non_exhaustive`